### PR TITLE
Add support for batched LU decomposition.

### LINKED ISSUE #655: Implement vmap/batching support for LU decomposition
I'm trying to solve linear equations Ax = b with vmap:

```python
import jax.numpy as np
import jax.random as random
from jax import vmap

key = random.PRNGKey(0)

A = random.normal(key, shape=(100, 2, 2))
x = random.normal(key, shape=(100, 2))
b = vmap(lambda A, x: A @ x)(A, x) # works! shape (100, 2)
x_hat = vmap(lambda A, x: np.linalg.solve(A, x))(A, x)
```

Here, using `np.linalg.solve` results in the following error message:

```
NotImplementedError: Batching rule for 'lu' not implemented
```

Opening this issue to request batching support for LU decomposition. Thanks in advance!

### PR DESCRIPTION
**Pull request by hawkinsp**
Fixes #655.

### CODE REVIEW

**Review by mattjj — APPROVED** [review]
Wow, got a lot of them at once!

**Comment by mattjj** [comment]
Do you want to add tests for the other batching rules (e.g. qr, svd) as well?

**Comment by hawkinsp** [comment]
I could test QR, and I'll do that. SVD doesn't actually have a batched implementation so I can't test it yet.

**Comment by hawkinsp** [comment]
It turned out I had to add some `core.pack()` logic to make the QR batching work, for reasons I don't entirely follow. PTAL.

**Review by mattjj — COMMENTED** [review]
Ah, the trickiness of primitives that return tuples. It's a blessing and a curse that there are so few examples. I also have to remind myself of this logic, so let's try to think through it.

Our general pattern is that Primitives (the Python-embedded representations of the primitives in jaxprs) only accept/return JaxTypes as inputs/outputs to their .bind() calls. JaxTypes are basically arrays and JaxTuples. For primitives that accept JaxTuples as arguments or produce them as outputs, our 'traceable' wrappers handle packing Python tuples into JaxTuples so that the caller needs not know about JaxTuples, and similarly handles unpacking JaxTuples into Python tuple outputs.

For example, for a Primitive `foo_p` and its corresponding traceable `foo`:

```python
def foo(xs):
  xs_jaxtuple = pack(xs)
  abc_jaxtuple = foo_p.bind(xs_jaxtuple)
  a, b, c = abc_jaxtuple
  return a, b, c
```

The invariant here is that the caller of the traceable doesn't see JaxTuples; they're hidden by this layer of indirection. (Traceables also handle the arg/kwarg conventions for calling Primitive.bind. That is, there are no constraints on how traceables are called in terms of args vs kwargs, but they canonicalize how the .bind() method is called.)

Transformation rules conceptually act on the primitive applications in jaxprs, which in effect means that on the Python-embedded side they act on the Primitive.bind calls rather than on traceables. Transformation rules should expect to see (and produce) JaxTuples in cases like these. Concretely, in this example, `qr_p.bind` produces a JaxTuple, and similarly its batching rule (like its JVP rule) should produce a JaxTuple as part of its output. That means we should expect to call `core.pack` on its output, or else just call `bind` directly if we want to avoid the traceable's unpacking. It's not a bad pattern for transformations to call .bind() rather than calling the traceable unless the traceable has some particular convenience logic we don't want to repeat. 

Does that make sense?

File: jax/lax_linalg.py [hunk 1]
```diff
@@ -47,8 +47,7 @@ def eigh(x, lower=True, symmetrize_input=True):
 def lu(x): return lu_p.bind(x)
 
 def qr(x, full_matrices=True):
-  q, r = qr_p.bind(x, full_matrices=full_matrices)
-  return q, r
+  return qr_p.bind(x, full_matrices=full_matrices)
```
[the remark this answers is not in the document]
Comment (hawkinsp) [reply jax/lax_linalg.py]: Done. And made the opposite change everywhere else in this file that returns a tuple (e.g. `eigh`)

File: jax/lax_linalg.py [hunk 2]
```diff
@@ -427,7 +426,8 @@ def qr_batching_rule(batched_args, batch_dims, full_matrices):
   x, = batched_args
   bd, = batch_dims
   x = batching.bdim_at_front(x, bd)
-  return qr(x, full_matrices=full_matrices), 0
+  q, r = qr(x, full_matrices=full_matrices)
+  return core.pack((q, r)), 0
```
[the remark this answers is not in the document]
Comment (hawkinsp) [reply jax/lax_linalg.py]: Done.

**Pull request merged**